### PR TITLE
Add url signing function to cloud

### PR DESCRIPTION
**Pull request by b123400**
connect #165

### CODE REVIEW

**Comment by rickmak** [comment]
This PR is fs only. I would prefer to merge with all 3 signer together. i.e. `fs`, `cloud` and `s3`

**Comment by b123400** [comment]
Updated to support s3, cloud and fs

File: lib/cloud/asset.js [hunk 1]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+    const host = this.settings.cloudAssetHost;
+    const url = `${host}/token/${appName}`;
+
+    return superagent.get(url)
+      .accept('application/json')
+      .set('Authorization', `Bearer ${token}`)
+      .query({expired_at: expire.toString()})
+      .then(response => {
+        const body = response.body;
+        console.log('fuck', body);
```
Comment (rickmak) [inline lib/cloud/asset.js]: The choice of word is too strong.

File: lib/cloud/settings.js [hunk 2]
```diff
@@ -78,5 +78,29 @@ export const settings = {
...
+  assetStoreURLPrefix: process.env.ASSET_STORE_URL_PREFIX ||
+    'http://127.0.0.1:3000/files',
+  assetStoreURLExpireDuration:
+    parseInt(process.env.ASSET_STORE_URL_EXPIRE_DURATION) || 3600,
+
+  // S3 asset settings
+  assetStoreS3AccessKey: process.env.ASSET_STORE_ACCESS_KEY || '',
+  assetStoreS3SecretKey: process.env.ASSET_STORE_SECRET_KEY || 'asset-secret',
+  assetStoreS3Bucket: process.env.ASSET_STORE_BUCKET || 'bucket-name',
+  assetStoreS3Region: process.env.ASSET_STORE_REGION || 'eu-west-1',
```
Comment (rickmak) [inline lib/cloud/settings.js]: Although we should not relay on default. I think the common default is `us-east-1`?

File: lib/cloud/asset.js [hunk 3]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
     body: data
   });
 }
+
+class Signer {
+  sign() {
```
Comment (rickmak) [inline lib/cloud/asset.js]: The sign interface is designed to return a `Promise`? May be wee some comment on this.

File: lib/cloud/asset.js [hunk 4]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+      .update(appName)
+      .update(name)
+      .update(expired.toString())
+      .update(this.signerToken.extra)
+      .digest('base64');
+    const signature_and_extra =
+      encodeURIComponent(`${hash}.${this.signerToken.extra}`);
+
+    return `${prefix}/${appName}/${name}` +
+      `?expired_at=${expired}&signature=${signature_and_extra}`;
```
Comment (rickmak) [inline lib/cloud/asset.js]: This return should be `Promise`?

File: lib/cloud/asset.js [hunk 5]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+    const url = this.signer.getUrl('GET', name, this.bucket, duration);
+    return Promise.resolve(url);
+  }
+}
+/**
+ * Return a signer for the current configuration.
+ *
+ * @return {Signer}
+ */
+export function getSigner() {
```
Comment (rickmak) [inline lib/cloud/asset.js]: We only need one instance of signer. The name `getSigner` also imply this to me.

So I have two quesiton:
- What is the consideration on returning new instance on every `getSigner`?
- The name `signerFactory` will tell everyone that the new signer will return on every call.

File: lib/cloud/asset.js [hunk 6]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+    super();
+    this.settings = _settings;
+    this.signerToken = null;
+  }
+
+  refreshSignerToken() {
+    const appName = this.settings.appName;
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const expire = Math.floor(Date.now() / 1000) + duration;
+    const token = settings.cloudAssetToken;
```
Comment (rickmak) [inline lib/cloud/asset.js]: `this.settings`

File: lib/cloud/asset.js [hunk 7]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+      .set('Authorization', `Bearer ${token}`)
+      .query({expired_at: expire.toString()})
+      .then(response => {
+        const body = response.body;
+        console.log('fuck', body);
+        this.signerToken = {
+          value: body.value,
+          expiredAt: Date.parse(body.expired_at),
+          extra: body.extra
+        };
```
Comment (rickmak) [inline lib/cloud/asset.js]: Just put those data in `this`, not `this.signerToken`? I think an object should not need two/three level of internal data structure to make things complex.

And flat structure it save typing ;)

File: lib/cloud/asset.js [hunk 8]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+      `?expired_at=${expired}&signature=${signature_and_extra}`;
+  }
+}
+
+class S3Signer extends Signer {
+  constructor(_settings) {
+    super();
+    this.settings = _settings;
+    const key = _settings.assetStoreS3AccessKey;
+    const secret = _settings.assetStoreS3SecretKey;
```
Comment (rickmak) [inline lib/cloud/asset.js]: Why not `this.settings`?

File: lib/cloud/asset.js [hunk 9]
```diff
@@ -47,3 +53,150 @@ export function staticAssetHandler(req) {
...
+  constructor(_settings) {
+    super();
+    this.settings = _settings;
+  }
+
+  sign(name) {
+    const prefix = this.settings.assetStoreURLPrefix;
+    const duration = this.settings.assetStoreURLExpireDuration;
+    const expire = Math.floor(Date.now() / 1000) + duration;
+    const secret = this.settings.assetStoreSecret;
```
Comment (rickmak) [inline lib/cloud/asset.js]: If we assume settings won't change. Just copy the setting into the `this`, not querying everytime during sign?

File: lib/cloud/asset.js (hunk 1 above)
Comment (b123400) [inline lib/cloud/asset.js]: Sorry this is totally debug code!!

**Pull request merged**